[KEP] Concurrent Admission (#8861)
kubernetes-sigs/kueue: keps/8691-concurrent-admission/README.md

      - [Multiple Preemptions](#multiple-preemptions)
      - [Reserving Quota for the same Workload twice](#reserving-quota-for-the-same-workload-twice)
      - [Scheduling Options in wrong order](#scheduling-options-in-wrong-order)
      - [FlavorFungibility](#flavorfungibility)
    - [Test Plan](#test-plan)
        - [Prerequisite testing updates](#prerequisite-testing-updates)
      - [Unit Tests](#unit-tests)
      - [Integration tests](#integration-tests)
    - [Graduation Criteria](#graduation-criteria)
      - [Alpha](#alpha)
      - [Beta](#beta)
  - [Implementation History](#implementation-history)
  - [Drawbacks](#drawbacks)
    - [Increased API Object Count](#increased-api-object-count)
  - [Alternatives](#alternatives)
    - [Modify AdmissionCheck API](#modify-admissioncheck-api)
    - [Migration within a Single Workload](#migration-within-a-single-workload)
<!-- /toc -->

## Summary

Currently, the Kueue admission process selects a single ResourceFlavor (RF) and pursues it until the workload is admitted. This KEP proposes Concurrent Admission, allowing a Workload to attempt multiple ResourceFlavors simultaneously.

> Kueue will create clones of a Workload, referred to as Options. Each Option is scheduled independently on a specific subset of ResourceFlavors. This unblocks scenarios where a user needs

Author: The lifecycle of each Option depends on the chosen `OnSuccessPolicy`, which I've tried to describe in the OnSuccess Policy section. Is there anything else you would like me to clarify?

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals
- [G1] Pursue capacity on multiple flavors (e.g. Spot and Reservation) simultaneously.

- [G2] Start two or more independent long-running AdmissionChecks until any succeeds.

- [G3] Support various policies on migration to more desired flavor

  - [a] Do not migrate at all

  - [b] Migrate only to "higher" RFs than the admitted one

  - [c] Migrate only to a subset of RFs

- [G4] Delay fallback to "lower" RFs than the currently tried

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> ### User Stories

Author: Yes, initially I imagined this feature to be entirely admin-facing.

> This could be problematic, especially for long-running workloads where disruptive migration may be too costly, for example migrating a multi-day or multi-week job that is near completion.

For such scenario we've introduced the `DeleteDelaySeconds` field which can limit the time when the migration can happen. I've added a story to clarify that.

> On the other hand, should we consider exposing any workload-level controls to opt in or out of this behavior?

If this feature is disabled on CQ level, I imagine opting in would be a bit tricky because there would be no configuration provided by a user, and we would need to use some default.

Opting out with some label/annotation could be easy. I could also imagine that some Workloads might want to use different configuration than others. However, I'm a bit hesitant to introduce additional user-facing API. A regular user would need to understand this feature, in particular they would need to understand  Kueue itself first. I think it should be an admin's role to set up cluster in the most optimal way for their users, but I'd love to hear if you know any use-cases, where users would like to have more control over a feature like this

To achieve that, I configure my ClusterQueue to use the Concurrent Admission with `ExplicitOptions` policy, where I set `CreateDelaySeconds` for the On-Demand option (details below).

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### ClusterQueue API
The ClusterQueue is extended to define the policy for concurrent attempts. This includes how to handle sibling Options once one is admitted and how to define specific, customized Options.

```
type OnSuccessPolicy string

const (
    // Stop all other attempts - no workload placement upgrade performed.
    RemoveOther OnSuccessPolicy = "RemoveOther"

    // Stop all attempts below the used RF - upgrade to higher RF possible.
    RemoveLower OnSuccessPolicy = "RemoveLower"

    // Stop all attempts below a defined target RF.
    RemoveBelowTarget OnSuccessPolicy = "RemoveBelowTarget"
)

> type ConcurrentAdmission struct {

Author: Done

**Default RF Options**: When Options are created automatically for each ResourceFlavor in the ClusterQueue, the name follows:

```
${original_workload_name}-option-${resource_flavor_name}
```

**Explicit Options**: When specific names are provided in the ExplicitOptions configuration:

```
${original_workload_name}-option-${explicit_option_name}
```

Note: Option names are designed to be deterministic. If a name collision occurs (due to long Workload/RF names), standard Kubernetes suffix truncation logic will be applied while maintaining the -option- identifier.

#### Workload Spec
```
type WorkloadType string
const (
    Default              WorkloadType = "Default"
    ResourceFlavorOption WorkloadType = "ResourceFlavorOption"
    Parent               WorkloadType = "Parent"
)

type WorkloadMode struct {
    // +required
    Type                   WorkloadType
}

type WorkloadSpec struct {
    ...

    // +optional
    WorkloadMode *WorkloadMode
}

type PodSet struct {
  ...

	// If set, only RF from this list can be assigned to PodSets.
  //
  // +optional

>   AllowedResourceFlavors []string

Author: I like the idea of embedding allowedResourceFlavors into the `schedulingConstraints`. When it comes to the other fields in this struct I'd like to continue the discussion in the other KEP. One thing to consider is that here we operate on PodSet level, and in the #8844 on the Workload level

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### ClusterQueue API
The ClusterQueue is extended to define the policy for concurrent attempts. This includes how to handle sibling Options once one is admitted and how to define specific, customized Options.

```
type ClusterQueueSpec struct {
    ...
    // +optional
    ConcurrentAdmission *ConcurrentAdmission
}

type OnSuccessPolicy string

const (
    // Stop all other attempts - no workload placement upgrade performed.
    RemoveOther OnSuccessPolicy = "RemoveOther"

    // Stop all attempts below the used RF - upgrade to higher RF possible.
    RemoveLower OnSuccessPolicy = "RemoveLower"

    // Stop all attempts below a defined target RF.
    RemoveBelowTarget OnSuccessPolicy = "RemoveBelowTarget"
)

type ConcurrentAdmission struct {
    // OnSuccess defines the policy applied when one of the options is admitted.
    //
    // +required
    OnSuccess OnSuccessPolicy

    // ExplicitOptions allows for fine-grained control over which options are created.
    // If not specified, Kueue creates an option for each RF mentioned in the CQ.
    //
    // +optional

>     ExplicitOptions []OptionCreationCustomization

Author: Done

    // RemoveBelowTargetConfig provides configuration for the RemoveBelowTarget policy.
    //
    // +optional

>     RemoveBelowTargetConfig *RemoveBelowTargetConfig

Reviewer: nit, I would propose to name the type ConcurrentAdmissionRemoveBelowTargetConfig. Yes it is longer, but this is common practice to make sure there are no type confiflicts. Note that all types in Kueue for v1beta2 are in the same package. Type names aren't user facing so this is not an issue

  // +optional
  AllowedResourceFlavors []string

}

```

#### Workload Status (Observability)

To ensure the user can easily check which attempts are still active, the WorkloadStatus is extended to include an Options list.
This provides a central view of all virtual workloads without needing to query for child objects manually.

For more detailed scheduling stats a user can check a particular Option Workload.

```
type WorkloadOptionStatus struct {
    // Name of the option (corresponds to the virtual workload name).
    Name string
    // ResourceFlavors assigned to this option.
    ResourceFlavors []string
    // Status of the admission for this specific option (e.g., Pending, Admitted, Evicted).
    Status string
    // Message provides details on why an option is in its current state.
    Message string
}

type WorkloadStatus struct {
    // ... existing fields ...

    // Options tracks the state of all concurrent admission attempts.
    Options []WorkloadOptionStatus
}
```

### Option Controller

A dedicated Option Lifecycle will be introduced to manage the state and lifespan of virtual Workloads and the relationship with the parent Workload.
Its responsibilities include:

> 1. Creation: It instantiates Options based on the CQ's ExplicitOptions or default RF list.

Reviewer 2: One clarification question around the Option Lifecycle Controller: when Option creation happens, is it performed in a standalone asynchronous reconciliation loop, or is it tied to the Workload’s first scheduling evaluation (i.e., during the initial scheduler pass)?

In other words, are Options expected to be created eagerly by the controller as soon as the Parent Workload is observed, or lazily as part of the scheduling/admission flow? Clarifying this would help reason about ordering, timing, and potential interactions with scheduling and AdmissionChecks.

If the former (i.e., Options are created in a standalone, asynchronous reconciliation loop), does the Option Lifecycle Controller evaluate or reason about ResourceFlavor assignment outside of the scheduler’s admission loop?

More specifically, does the controller make any decisions about which flavors an Option should target beyond applying static constraints (e.g., `AllowedResourceFlavors`), or is all flavor selection and feasibility still exclusively determined by the scheduler during admission? 

Does the Option Lifecycle Controller create Options for all ResourceFlavors defined in the ClusterQueue (or ExplicitOptions), even if some of those flavors cannot currently accommodate the Workload due to obvious constraints (e.g. insufficient quota, incompatible topology, AdmissionChecks that are known to fail)?

Or is there any upfront filtering at Option creation time to avoid creating Options for flavors that are clearly infeasible, with all feasibility checks deferred entirely to the scheduler? Clarifying this would help understand whether Option creation is purely declarative or includes any early feasibility evaluation.

    // Name of the option (corresponds to the virtual workload name).
    Name string
    // ResourceFlavors assigned to this option.
    ResourceFlavors []string
    // Status of the admission for this specific option (e.g., Pending, Admitted, Evicted).
    Status string
    // Message provides details on why an option is in its current state.
    Message string
}

type WorkloadStatus struct {
    // ... existing fields ...

    // Options tracks the state of all concurrent admission attempts.
    Options []WorkloadOptionStatus
}
```

### Option Controller

A dedicated Option Lifecycle will be introduced to manage the state and lifespan of virtual Workloads and the relationship with the parent Workload.
Its responsibilities include:

1. Creation: It instantiates Options based on the CQ's ExplicitOptions or default RF list.
2. Aggregation: It syncs the status of Options back into the status.options of the Parent. Based on that the Parent Workload can interact (e.g. suspend/unsuspend) with the top-level job
3. Policy Enforcement: Upon admission of any option, it executes the OnSuccessPolicy (e.g., deactivating lower-priority virtual Workloads).

### Observability

> Users can track the progress of concurrent attempts via:

Reviewer: How are the metrics going to work? For example we have metrics counting the "pedning" workload, admitted , finished workloads, etc. Are they going to be incremented per option, per Workload (parent + option), or just by parent?

Status: The Parent Workload status will list status of all Options.

Checking a particular Option directly.

### Notes/Constraints/Caveats (Optional)

#### Code Changes Complexity

The feature should require as few changes in the quota accounting, scheduling, other core Kueue features as possible.
It creates Workloads that should be treated as regular Workloads, but are controlled by the `OptionsController` that
can create/activate/deactivate Options.

With making as little changes to scheduling logic as possible in mind, we still need to work on at least 3 things there:
1) Narrowing selection of ResourceFlavors for a given Workload. This however can be also used outside of the Concurrent Admissions feature, creating more flexibility for Kueue.
2) Preempting sibling Options when admitting more preferable ones.

> ### Risks and Mitigations

Reviewer: One risk to me is debuggability for the time-based options. It is sometimes hard to know when the accounting period started, etc.

#### StrictFIFO

Handling `StrictFIFO` queueing strategy comes with challenges.

The most important one is once less favorable Options is admitted,
scheduler cannot try admitting more favorable Options (migration)
and let other Workloads schedule at the same time, without violating `StrictFIFO` semantics.
It can either try the more favorable Option over and over which blocks the queue, or
let other Workloads schedule, which violates the `StrictFIFO`

The other one is to not block Options siblings if the first one cannot be scheduled. There are at least couple of ways to solve it:

1) Grouped Popping - the scheduler could pick up all of the Options belonging to the same parent from the heap and then process them within the same schedulingCycle. Each heap node would still consist of one Workload.
2) The heap node consists of all of the Options belonging to the same parent Workload.
3) The scheduler picks up only one Option per scheduling cycle. If there's a sibling in the heap head, it doesn't put back the processed after failed scheduling. Instead it lets the other Option run.

> Mitigation: For Alpha and Beta version of this feature we don't plan to support `StrictFIFO` queueing strategy. Based on users' feedback we will reconsider it for GA.

Reviewer: Let's don't put it into the risk section: https://github.com/kubernetes-sigs/kueue/blob/d07bd880128ed9688bca92a6251f05c58b597bcf/keps/NNNN-template/README.md?plain=1#L120-L132

Risks are mostly for things we implement, but foresee as risks due to integration with other components etc. Complexity of the implementation is typically not considered a risk

Author: Got it, moved to design details section

#### Reserving Quota for the same Workload twice

Migrating to a more preferred flavor could lead to booking quota for the same Workload twice — once for the running instance and once for the "upgrade" attempt.

Mitigation: The Kueue scheduler must preempt the less preferable Option immediately before admitting the more preferable one to ensure resource utilization remains accurate and fair sharing is not negatively impacted.

#### Scheduling Options in wrong order

The `OptionsController` leverages the existing implementation of scheduling logic in Kueue, where Workloads are sorted
by priorities and creation timestamps and then picked-up by the scheduler one per scheduling cycle (per ClusterQueue).
This means when creating Options the controller should ensure more favorable Options are ahead of less favorable ones in the queue. This could be achieved by adding another dimension to our heap sort mechanism that would only
be used for sorting Options coming from the same Parent. The value of this dimension would be filled by the Option Controller based on the order of ResourceFlavors in ClusterQueue.

#### FlavorFungibility

> In the first iteration of the feature we don't plan to integrate with the FlavorFungibility configuration. Users should be able to

Author: After sync offline, expanded even more and moved back to Risks section

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals
- Pursue capacity on multiple flavors (e.g. Spot and Reservation) simultaneously.

- Start two or more independent long-running AdmissionChecks until any succeeds.

- Support various policies on migration to more desired flavor

  - Do not migrate at all

  - Migrate only to "higher" RFs than the admitted one

  - Migrate only to a subset of RFs

  - Do not migrate after certain amount of time

- Delay fallback to "lower" RFs than the currently tried

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

- Describe the exact behavior of ConcurrentAdmission with `StrictFIFO` queueing strategy.

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

Concurrent Admission feature propose to introduce a new controller and extend the ClusterQueue API.
Based on the API a new controller creates multiple Option Workloads per a single top-level job.
At any given point in time, only one Option can be admitted by Kueue.

> Each Option is a clone of Parent Workload with additional scheduling constraints, in particular could only be assigned to a subset of ResourceFlavors

Reviewer 2: Will this still be true for a single-RF setup, i.e. Job → Workload → Workload (Option)?

My intuition is that it shouldn’t, since there is effectively no choice to make. Either way, it would be helpful to call this out explicitly in the KEP so readers understand whether the option machinery is always present (even for a single candidate) or only created when there are multiple candidates.

To achieve that, I configure my ClusterQueue to use the Concurrent Admission feature, with `RemoveOther` policy (details below).

#### Story 5: Delaying Option creation
As an admin I have two resource flavors in my cluster:
1) Most preferable: Reservation
2) Less preferable: On-Demand

I want my workloads to attempt scheduling on 'Reservation' only for the first 2 hours. If they are not admitted, I want to try 'Reservation' and 'On-Demand' simultaneously.

To achieve that, I configure my ClusterQueue to use the Concurrent Admission with `ExplicitOptions` policy, where I set `CreateDelaySeconds=7200` for the On-Demand option (details below).

#### Story 6: Limit when migration can happen
As an admin I have two resource flavors in my CQ:
1) Most preferable: Reservation
2) Less preferable: On-Demand

I know at some point I'm going to run a long-running Workload with high migration cost. I want to avoid migration near its completion
and hence I want to constraint it to happen only within the first day.

> To achieve that, I configure my ClusterQueue to use the Concurrent Admission with `ExplicitOptions` policy, where I set `DeleteDelaySeconds=86400` for the Reservation option (details below).

Reviewer 2: I think this story **partially addresses** my earlier concern, but it also highlights an important limitation of the proposed approach.

The key issue with `DeleteDelaySeconds` is not the specific threshold value itself, but the **nature of the threshold**. Using a fixed duration expressed in seconds assumes that all workloads submitted to a given ClusterQueue have roughly similar lifetimes. In practice, that is often not the case, workloads can vary widely in duration.

As a result, a single fixed value may be:

* too short for long-running workloads with high migration cost, or
* too long for short-lived workloads where migration late in execution is unnecessary or harmful.

Have you considered an alternative that is **relative rather than absolute**, for example a percentage-based threshold tied to the workload’s declared or expected duration? This could better align migration decisions with workload progress rather than wall-clock time.

This is very much an off-the-top idea and I’m not sure whether it is feasible as described, but I wanted to call out the limitation of a fixed “seconds-based” approach explicitly, since it may not generalize well across heterogeneous workloads.

#### Story 7: Workload with multiple PodSets
As an admin I have two GPU resource flavors and one CPU resource flavor in my cluster.
GPU:
1) Most preferable: Reservation
2) Less preferable: On-Demand

CPU:
1) Default-CPU

My workload consists of 2 PodSets, one requesting GPUs and one requesting CPUs only.
I want my workloads to start as soon as possible, on whatever GPU flavor. However if the Reservation flavor releases some quota, I want to migrate to that flavor.

To achieve that, I configure my ClusterQueue to use the Concurrent Admission with `ExplicitOptions` policy.
I create a configuration for the Reservation Option with `AllowedResourceFlavors=["Reservation, Default-CPU"]`
and for the On-Demand Option with `AllowedResourceFlavors=["On-Demand", "Default-CPU"]`.

> 

Author: Still, I'd argue the worst case scenario depends on the use-case. With Concurrent Admission the number of Options per Job is a product of all migration dimension. If the only dimension we want to migrate is GPU, and we treat different CPUs flavors as equal good, then the product is equal 1 (CPU dimension) x #GPU-flavors. If we want to migrate both in GPU and CPU dimensions then indeed, the product is #CPU-flavors x #GPU-flavors. 

Theoretically, this API allows to create 2^#RF-2 different Options, because that's the number of all RF subsets excluding an empty one, and the one containing all RFs. However in environment with a lot of different flavors I treat it more as a misconfiguration, rather than the real-world  worst case scenario. Misconfiguration is already one the points in the Risk section

The most important one is once less favorable Options is admitted,
scheduler cannot try admitting more favorable Options (migration)
and let other Workloads schedule at the same time, without violating `StrictFIFO` semantics.
It can either try the more favorable Option over and over which blocks the queue, or
let other Workloads schedule, which violates the `StrictFIFO`

The other one is to not block Options siblings if the first one cannot be scheduled. There are at least couple of ways to solve it:

1) Grouped Popping - the scheduler could pick up all of the Options belonging to the same parent from the heap and then process them within the same schedulingCycle. Each heap node would still consist of one Workload.
2) The heap node consists of all of the Options belonging to the same parent Workload.
3) The scheduler picks up only one Option per scheduling cycle. If there's a sibling in the heap head, it doesn't put back the processed after failed scheduling. Instead it lets the other Option run.

For Alpha and Beta version of this feature we don't plan to support `StrictFIFO` queueing strategy. Based on users' feedback we will reconsider it for GA.

> ### Risks and Mitigations

Author: Explained it in the Drawback section and mentioned it in Motivation section as well, ptal

### Scheduling & Lifecycle
An Option Workload functions near-identically to a "regular" Workload regarding quota accounting, preemption, and core scheduling features.

At any given point in time, only one Option per Parent may be admitted by Kueue.

To support this, we will introduce a new controller and extend the ClusterQueue API with a new `.spec` field to manage Option activation and deactivation.
<!-- 
Those new types of Workloads can freely coexist with "regular" Workloads in a cluster.

The relationship between a Parent and its Options is parent–child, with 1:1+ cardinality.
Every Parent has at least one Option, and potentially more depending on PodSets and ClusterQueue ResourceFlavors.
All of them are separate k8s objects stored in etcd.

A parent Workload is excluded from the scheduling logic in Kueue. It acts as a owner and status aggregator of Options.

An Option Workload is a cloned view of its Parent with some additional scheduling constraints,
in particular it can be scheduled on a limited number of ResourceFlavors.

Apart from that an Options Workload acts almost identically as a "regular" Workload regarding scheduling, quota accounting and other core features.

> At any given point in time, only one Option can be admitted by Kueue.

Reviewer 2: What is the mechanism for Option succession when one Option replaces another during flavor migration?
It would probably be a good idea to describe this explicitly in the KEP.

```
type ClusterQueueSpec struct {
    ...
    // +optional
    ConcurrentAdmission *ConcurrentAdmission
}

type OnSuccessPolicy string

const (
    // Stop all other attempts - no workload placement upgrade performed.
    RemoveOther OnSuccessPolicy = "RemoveOther"

    // Stop all attempts below the used RF - upgrade to higher RF possible.
    RemoveLower OnSuccessPolicy = "RemoveLower"

    // Stop all attempts below a defined target RF.
    RemoveBelowTarget OnSuccessPolicy = "RemoveBelowTarget"
)

type ConcurrentAdmission struct {
    // OnSuccess defines the policy applied when one of the options is admitted.
    //
    // +required
    OnSuccess OnSuccessPolicy

    // ExplicitOptions allows for fine-grained control over which options are created.
    // If not specified, Kueue creates an option for each RF mentioned in the CQ.
    //
    // +optional
    ExplicitOptions []ConcurrentAdmissionExplicitOption

    // RemoveBelowTargetConfig provides configuration for the RemoveBelowTarget policy.
    //
    // +optional
    RemoveBelowTargetConfig *ConcurrentAdmissionRemoveBelowTargetConfig
}

> type OptionCreationCustomization struct {

Author: Done

### Workload API

An Option Workload references its Parent in the `metadata.ownerReferences` field.
Besides that a Parent Workload has the `kueue.x-k8s.io/parent-option` annotation to distinguish it from
"regular" Workloads, in a cluster.

This way of distinguishing Options, Parents, and "regular" Workloads is introduced in the Alpha, and will be
revisited when graduating to Beta.

#### Naming Convention

To maintain a clear relationship between the parent and its virtual clones, the Option Workloads will follow a strict naming convention.

**Default RF Options**: When Options are created automatically for each ResourceFlavor in the ClusterQueue, the name follows:

```
${original_workload_name}-option-${resource_flavor_name}
```

**Explicit Options**: When specific names are provided in the ExplicitOptions configuration:

```
${original_workload_name}-option-${explicit_option_name}
```

> Note: Option names are designed to be deterministic. If a name collision occurs (due to long Workload/RF names), standard Kubernetes suffix truncation logic will be applied while maintaining the -option- identifier.

Author: Sorry, I meant the Kueue logic, updated

#### Workload Spec

We propose to extend Workload's Spec API to have the capability to narrow the set of ResourceFlavors that can be used
for scheduling a Workload.

```
type WorkloadSpec struct {
    ...
    // AdmissionConstraints describes the constraints Kueue scheduling algorithm takes into account
    // when reserving quota for a Workload.
    //
    // +optional
    AdmissionConstraints *AdmissionConstraints
}

type AdmissionConstraints struct {
  ...

	// If set, only RF from this list can be assigned to this Workload.
  //
  // +optional
  AllowedResourceFlavors []ResourceFlavorReference
}

```

#### Workload Status

To ensure the user can easily check which attempts are still active, the WorkloadStatus is extended to include an Options list.
This provides a central view of all virtual workloads without needing to query for child objects manually.

For more detailed scheduling stats a user can check a particular Option Workload.

```
type WorkloadOptionState string

const (
    // OptionsStatePending means the Option is still pending
    OptionStatePending = "Pending"

    // OptionStateAdmitted means the Option has been admitted

>     OptionStateAdmitted = "Admitted"

Reviewer: Couldn't the option mechanism be used with AdmissinoChecks? If so then we distinguish also the "QuotaReserved" state. Does the design consider them as "Pending"? I'm not clear if we need to make that distinction, but this information seems useful for decision making.

to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

#### Unit Tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `<package>`: `<date>` - `<test coverage>`

#### Integration tests

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

### Graduation Criteria

> #### Alpha

Author: Sure, we can gate it for Alpha

Initial support for `RemoveBelowTarget` policy.

Integration with `BestEffortFIFO` queueing strategy.

Introduction of `WorkloadOptionStatus` and `AdmissionConstraints` fields.

#### Beta

Support for `RemoveOther` and `RemoveLower` policies.

Introduction of `ExplicitOptions` functionality.

Revisit extending `ExplicitOptions` API with some additional fields.

Minimizing number of Options issuing preemptions to only one per Parent.

Revisit the idea of introducing WorkloadType API
```
type WorkloadType string
const (
    Default              WorkloadType = "Default"
    ResourceFlavorOption WorkloadType = "ResourceFlavorOption"
    Parent               WorkloadType = "Parent"
    ... // possibly more like WorkloadSlice, PrebuiltWorkload

> )

Author: Done

3) Least preferable: Spot

I want my workloads to start as soon as possible, on whatever flavor. However if more preferable flavor releases some quota, I want to migrate to that flavor.

e.g. Workload is running on Spot, but after 1h On-Demand quota is released so I want to use it instead.
e.g. Workload is running on On-Demand, but after 1h Reservation quota is released so I want to use it instead.

To achieve that, I configure my ClusterQueue to use the Concurrent Admission feature, with `UpgradeAboveCurrent` policy (details below).

#### Story 2: Upgrade only to Reservation
As an admin I have three resource flavors in my cluster:
1) Most preferable: Reservation
2) Less preferable: On-Demand
3) Least preferable: Spot

I want my workloads to start as soon as possible, on whatever flavor. However if more Reservation releases some quota, I want to migrate to it.

e.g. Workload is running on Spot, but after 1h On-Demand quota is released. I don't want to migrate to that flavor because the cost of migration surpasses
the possible gain of migration.

e.g. Workload is running on On-Demand, but after 1h Reservation quota is released so I want to use it instead.

> To achieve that, I configure my ClusterQueue to use the Concurrent Admission feature, with `UpgradeAboveTargetTarget: Reservation` policy (details below).

Reviewer: nit, fix typo "targetTarget", here and in other places

Author: Done

### Workload API

An Option Workload references its Parent in the `metadata.ownerReferences` field.
Besides that a Parent Workload has the `kueue.x-k8s.io/parent-option` annotation to distinguish it from
"regular" Workloads, in a cluster.

This way of distinguishing Options, Parents, and "regular" Workloads is introduced in the Alpha, and will be
revisited when graduating to Beta.

#### Naming Convention

To maintain a clear relationship between the parent and its virtual clones, the Option Workloads will follow a strict naming convention.

**Default RF Options**: When Options are created automatically for each ResourceFlavor in the ClusterQueue, the name follows:

```
${original_workload_name}-option-${resource_flavor_name}
```

**Explicit Options**: When specific names are provided in the ExplicitOptions configuration:

```
${original_workload_name}-option-${explicit_option_name}
```

Note: Option names are designed to be deterministic. If a name collision occurs (due to long Workload/RF names), standard Kueue suffix truncation logic will be applied while maintaining the -option- identifier.

> #### Workload Spec

Reviewer 3: Is there any way to represent such a preference in the top-level Job object?
As I check this proposal, those preferences could be represented only at the ClusterQueue level, right?

}

type ConcurrentAdmission struct {
    // MigrationConstraints defines the constraints of Variants' migration
    //
    // +required
    MigrationConstraints ConcurrentAdmissionMigrationConstraints

    // ExplicitVariants allows for fine-grained control over which variants are created.
    // If not specified, Kueue creates a variant for each RF mentioned in the CQ.
    //
    // +optional
    // +kubebuilder:validation:MaxItems=16
    ExplicitVariants []ConcurrentAdmissionExplicitVariant
}

type ConcurrentAdmissionMigrationMode string

const (
    // Do not allow any kind of migration
    NoMigration ConcurrentAdmissionMigrationMode = "NoMigration"

    // Allow upgrades
    UpgradeOnly ConcurrentAdmissionMigrationMode = "UpgradeOnly"
)

type ConcurrentAdmissionMigrationConstraints struct {
    // Mode defines the mode of Workload's migration.
    //
    // +required
    Mode ConcurrentAdmissionMigrationMode

    // MinFlavor defines the minimal flavor a Workload can migrate to.
    // The order is based on the order of flavors in ClusterQueue.
    // It can only be used if the Mode is `UpgradeOnly` and `ExplicitVariants` is not specified.

>     // It the Mode is `UpgradeOnly` and MinFlavor is not specified, then there's

Author: Done

    // MinVariant defines the minimal Variant a Workload can migrate to.
    // The order is based on the order of variant in `ExplicitVariants`.
    // It can only be used if the Mode is `UpgradeOnly` and `ExplicitVariants` is specified.
    // It the Mode is `UpgradeOnly` and MinVariant is not specified, then there's
    // no constraints on what Variant a Workload can migrate to.
    //
    // +optional
    MinVariant *string
}

type ConcurrentAdmissionExplicitVariant struct {
    // Name of the variant, must be unique within the ClusterQueue.
    //
    // +required
    Name string

    // CreateDelaySeconds defines how long after Workload creation this Variant is activated.
    // Allows prioritizing a preferred RF before falling back to others.
    //
    // +optional
    CreateDelaySeconds *int32

    // DeleteDelaySeconds defines how long after admission of other Variants,
    // this Variant should be deactivated.
    // Allows disallowing migration after certain amount of time
    //
    // +optional
    DeleteDelaySeconds *int32

    // AllowedResourceFlavors limits which flavors can be assigned to PodSets for this variant.
    //
    // +required

>     AllowedResourceFlavors []ResourceFlavorReference

Reviewer 4: When there are multiple ResourceGroups, will you validate that this list contains at least one RF from each?
(Otherwise such Variant is not useful... except for cases when a workload omits requests for some of the CQ's resources. So maybe because of this case we shouldn't have such validation. WDYT?)

```

Note: Variant names are designed to be deterministic. If a name collision occurs (due to long Workload/RF names), standard Kueue suffix truncation logic will be applied while maintaining the -variant- identifier.

#### Workload Spec

We propose to extend Workload's Spec API to have the capability to narrow the set of ResourceFlavors that can be used
for scheduling a Workload.

```
type WorkloadSpec struct {
    ...
    // AdmissionConstraints describes the constraints Kueue scheduling algorithm takes into account
    // when reserving quota for a Workload.
    //
    // +optional
    AdmissionConstraints *AdmissionConstraints
}

type AdmissionConstraints struct {
  ...

	// If set, only RF from this list can be assigned to this Workload.
  //
  // +optional
  AllowedResourceFlavors []ResourceFlavorReference
}

```

#### Workload Status

To ensure the user can easily check which attempts are still active, the WorkloadStatus is extended to include a Variants list.
This provides a central view of all virtual workloads without needing to query for child objects manually.

For more detailed scheduling stats a user can check a particular Variant Workload.

```

> type WorkloadVariantStatus struct {

Reviewer: I think we discussed the Workload OptionStatus is not necessarily needed for Alpha, and we could track the status in the cache for now. Could we move it to alternatives and re-evaluate in the next releases?

    Variants []WorkloadVariantStatus
}
```

#### StrictFIFO

Handling `StrictFIFO` queueing strategy comes with challenges.

The most important one is once less favorable Variants is admitted,
scheduler cannot try admitting more favorable Variants (migration)
and let other Workloads schedule at the same time, without violating `StrictFIFO` semantics.
It can either try the more favorable Variant over and over which blocks the queue, or
let other Workloads schedule, which violates the `StrictFIFO`

The other one is to not block Variants siblings if the first one cannot be scheduled. There are at least couple of ways to solve it:

1) Grouped Popping - the scheduler could pick up all of the Variants belonging to the same parent from the heap and then process them within the same schedulingCycle. Each heap node would still consist of one Workload.
2) The heap node consists of all of the Variants belonging to the same parent Workload.
3) The scheduler picks up only one Variant per scheduling cycle. If there's a sibling in the heap head, it doesn't put back the processed after failed scheduling. Instead it lets the other Variant run.

> For Alpha and Beta version of this feature we don't plan to support `StrictFIFO` queueing strategy. Based on users' feedback we will reconsider it for GA.

Reviewer 4: "We don't plan to support" - does it mean "we'll validate against such combination" or "we'll allow it but treat it as 'do at your own risk, effects are unspecified' "?

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

#### Unit Tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

> - `<package>`: `<date>` - `<test coverage>`

Author: Updated

"regular" Workloads, in a cluster.

This way of distinguishing Variants, Parents, and "regular" Workloads is introduced in the Alpha, and will be
revisited when graduating to Beta.

#### Naming Convention

To maintain a clear relationship between the parent and its virtual clones, the Variant Workloads will follow a strict naming convention.

**Default RF Variants**: When Variants are created automatically for each ResourceFlavor in the ClusterQueue, the name follows:

```
${job-kind}-${job-name}-variant-${resource-flavor}-${hash}
```

Example:
```
job-abc-variant-reservation-foohash
```

**Explicit Variants**: When specific names are provided in the ExplicitVariants configuration:

```
${job-kind}-${job-name}-variant-${variant-name}-${hash}
```

Example:
```
job-abc-variant-bar-foohash
```

#### Workload Spec

We propose to extend Workload's Spec API to have the capability to narrow the set of ResourceFlavors that can be used
for scheduling a Workload.

```
type WorkloadSpec struct {
    ...

    // AdmissionConstraints describes the constraints Kueue scheduling algorithm takes into account
    // when reserving quota for a Workload.
    //
    // +optional

>     AdmissionConstraints *AdmissionConstraints

Reviewer 3: Does this admissionConstraints ignore other constraints like flavor taints and tolerations?

Author: It respects flavor taints and tolerations, added clarification

During the Alpha implementation phase we will evaluate if the mechanism is done based on in-memory tracking, or we use the preemptionGates mechanism [KEP 8303](/keps/8303-multikueue-orchestrated-preemption/README.md).

### Ordering Variants

The `VariantsController` leverages the existing implementation of scheduling logic in Kueue, where Workloads are sorted
by priorities and creation timestamps and then picked-up by the scheduler one per scheduling cycle (per ClusterQueue).
This means when creating Variants the controller should ensure more favorable Variants are ahead of less favorable ones in the queue.
This could be achieved by adding another dimension to our heap sort mechanism that would only
be used for sorting sibling Variants. The value of this dimension would be filled by the Variant Controller based on the order of ResourceFlavors in ClusterQueue.
Thanks to that we also have a guarantee that sibling Variants are always adjacent in the heap, which
results in Kueue scheduler picking sibling Variants one after the another, without any other Workloads in between.

### Notes/Constraints/Caveats (Optional)

> #### Beta API

Reviewer 3: Can you move these Beta APIs to alternative sections?

Author: Done

2) Evicting sibling Variants when admitting more preferable ones.

### Reserving Quota for the same Workload twice

Migrating to a more preferred flavor could lead to booking quota for the same Workload twice — once for the running instance and once for the "upgrade" attempt.
A more preferred Variant must evict the less preferable one immediately before admission to ensure resource utilization remains accurate, and fair sharing is not negatively impacted.

### Multiple Preemptions

When pursuing multiple flavors concurrently, Kueue might preempt Workloads to accommodate multiple Variants belonging to the same Parent Workload.
While we only issue preemptions coming from one Workload per CQ, what happens is:
1. A Variant preempted a Workload and got the quota reserved.
2. The same Variants is now running AdmissionChecks
3. A sibling Variant is picked up by the scheduler and is preempting some other Workloads

We want to disallow other Variants to preempt if one of the Variants has already preempted some Workloads.

> We achieve it by storing in memory a map of Workloads that have issued a preemption during their current admission cycle.

Reviewer 5: We did discuss it some time ago and while I think the initial decision was to do this in-memory in this KEP, the preemption gates were designed in a way that should support Concurrent Admission out-of-the-box we want to use them here now/in the future.

2) Less preferable: On-Demand
3) Least preferable: Spot

I want my workloads to start as soon as possible, on whatever flavor. However, if a more preferable flavor releases some quota, I want to migrate to that flavor.

To achieve that, I configure my ClusterQueue to use the UpgradeOnly mode.

```
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: UpgradeOnly
```

#### Story 2: Upgrade only to Reservation
As an admin I have three resource flavors in my cluster:
1) Most preferable: Reservation
2) Less preferable: On-Demand
3) Least preferable: Spot

I want my workloads to start as soon as possible, on whatever flavor. However, I only want to incur the cost of migration if it means landing on the "Reservation" flavor.

To achieve that, I set the minTargetFlavor to "reservation". This ensures that a workload on "Spot" will not migrate to "On-Demand," only to "Reservation."

```
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: UpgradeOnly

>       minTargetFlavor: "reservation"

Author: Since this is the name of the ResourceFlavor I believe it needs to be lower case to follow k8s conventions

>       minTargetFlavor: "reservation"

Author: Done!

```
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: UpgradeOnly
      minTargetFlavor: reservation
```

#### Story 4: Homogenous Flavors only
As an admin, I have three homogeneous resource flavors (1a, 1b, 1c). I want my workloads to start as soon as possible on any flavor and stop pursuing other variants once the job is accommodated.

To achieve that, I use the `NoMigration` mode.

```
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: NoMigration
```

#### Story 5: Delaying Variant creation
As an admin I have two resource flavors in my cluster:
1) Most preferable: Reservation
2) Less preferable: On-Demand

I want to attempt scheduling on "Reservation" exclusively for the first 2 hours before allowing the "On-Demand" variant to activate.

To achieve this, I use explicitVariants and set createDelaySeconds for the On-Demand variant.

> ```

Reviewer 3: ```suggestion
```yaml
```

nit

apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation"
        allowedResourceFlavors: ["reservation"]
      - name: "on-demand"
        allowedResourceFlavors: ["on-demand"]
        createDelaySeconds: 7200
```

#### Story 6: Limit when migration can happen
As an admin I have two resource flavors in my CQ:
1) Most preferable: Reservation
2) Less preferable: On-Demand

I want to avoid migrating a long-running workload if it is nearing completion. I want to constrain the migration window to the first 24 hours (86,400 seconds) of the workload's lifecycle.

To achieve this, I set maxDeleteDelaySeconds on the more preferred variant. After this time passes, the "reservation" variant is deactivated, preventing any further upgrades.

```
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation"

>         allowedResourceFlavors: ["reservation"]

Reviewer 3: ```suggestion
        allowedResourceFlavors: ["Reservation"]
```

        maxDeleteDelaySeconds: 86400
      - name: "on-demand"
        allowedResourceFlavors: ["on-demand"]
```

#### Story 7: Workload with multiple PodSets
As an admin I have two GPU resource flavors and one CPU resource flavor in my cluster.
GPU:
1) Most preferable: Reservation
2) Less preferable: On-Demand

CPU:
1) Default-CPU

I want to allow upgrades for the GPU portion while keeping the CPU flavor constant across variants.

```
apiVersion: kueue.x-k8s.io/v1beta2
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmission:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation-flavor"
        allowedResourceFlavors: ["reservation", "default-cpu"]
      - name: "on-demand-flavor"

>         allowedResourceFlavors: ["on-demand", "default-cpu"]

Reviewer 3: ```suggestion
        allowedResourceFlavors: ["On-Demand", "Default-CPU"]
```

```

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> ## Design Details

Author: For Alpha we won't allow using WorkloadSlices and ConcurrenteAdmission at the same time. I think we should wait until both features mature and we collect some feedback about potential use cases - right now we don't have

> ## Design Details

Reviewer 3: That sounds reasonable to me, let's adding WorkloadSlice support to graduation criteria as I commented in the following new comments.

In this section, we often refer to `ExplicitVariants`. This concept is explained in more detail in [the
Explicit Variants API section](#explicitvariants-api).

### ClusterQueue API
The ClusterQueue is extended to define the policy for concurrent attempts. This includes how to handle sibling Variants once one is admitted and how to define specific, customized Variants.

```
type ClusterQueueSpec struct {
    ...

    // +optional

>     ConcurrentAdmission *ConcurrentAdmission

Author: Changed as suggested, and added json tags

}

type ConcurrentAdmission struct {
    // MigrationConstraints defines the constraints of Variants' migration
    //
    // +required

>     MigrationConstraints ConcurrentAdmissionMigrationConstraints

Author: Added a section (#lifecycle-of-a-parent-and-its-variants) to describe it

>     MigrationConstraints ConcurrentAdmissionMigrationConstraints

Reviewer 3: Thank you for adding that!
Based on `Lifecycle of a Parent and its Variants` description, I am confident that the ConcurrentAdmission behavior is exactly migration. Especially, "or keep
trying to get scheduled based on `migrationConstraints` configuration."

}

type ConcurrentAdmissionMigrationConstraints struct {
    // Mode defines the mode of Workload's migration.
    //
    // +required
    Mode ConcurrentAdmissionMigrationMode

    // MinTargetFlavor defines the minimal flavor a Workload can migrate to.
    // The order is based on the order of flavors in ClusterQueue.
    // It can only be used if the Mode is `UpgradeOnly` and `ExplicitVariants` is not specified.
    // If the Mode is `UpgradeOnly` and MinTargetFlavor is not specified, then there's
    // no constraints on what flavors a Workload can migrate to.
    //
    // +optional
    MinTargetFlavor *ResourceFlavorReference
}

type ConcurrentAdmissionMigrationMode string

const (
    // Do not allow any kind of migration
    NoMigration ConcurrentAdmissionMigrationMode = "NoMigration"

    // Allow upgrades
    UpgradeOnly ConcurrentAdmissionMigrationMode = "UpgradeOnly"
)
```

In Alpha this API will only be supported in ClusterQueues that have 1 `ResourceGroup`. We will validate against greater number.
This will be reconsidered when graduating the feature to Beta.

### Workload API

> 

Author: We can definitely start with something stronger, 16 sounds good to me. I'll clarify that in the KEP, thanks

> 

Author: Added

A Variant Workload is a cloned view of its Parent with some additional scheduling constraints,
in particular it can be scheduled on a limited number of ResourceFlavors.

Apart from that a Variant Workload acts almost identically as a "regular" Workload regarding scheduling, quota accounting and other core features.

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories

#### Story 1: ResourceFlavor Upgrade
As an admin I have three resource flavors in my cluster:
1) Most preferable: reservation
2) Less preferable: on-demand
3) Least preferable: spot

I want my workloads to start as soon as possible, on whatever flavor. However, if a more preferable flavor releases some quota, I want to migrate to that flavor.

To achieve that, I configure my ClusterQueue to use the UpgradeOnly mode.

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

It seems that the code block starting marks are missing.

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
```

#### Story 2: Upgrade only to reservation
As an admin I have three resource flavors in my cluster:
1) Most preferable: reservation
2) Less preferable: on-demand
3) Least preferable: spot

I want my workloads to start as soon as possible, on whatever flavor. However, I only want to incur the cost of migration if it means landing on the "reservation" flavor.

To achieve that, I set the minTargetFlavor to "reservation". This ensures that a workload on "spot" will not migrate to "on-demand," only to "reservation."

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

ditto

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
      minTargetFlavor: "reservation"
```

#### Story 3: reservation + Homogenous Flavors
As an admin I have four resource flavors in my cluster:

1) Most preferable: reservation
2) Less preferable flavors with a long-running AdmissionCheck
   - A
   - B
   - C

I want my workloads to start as soon as possible. I want to migrate to "reservation" if it becomes available, but I don't want to migrate between the homogeneous 2a/2b/2c flavors.

By setting the minTargetFlavor to "Reservation", Kueue will ignore available capacity in 2a if the workload is already running on 2b.

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

ditto

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
      minTargetFlavor: reservation
```

#### Story 4: Homogenous Flavors only
As an admin, I have three homogeneous resource flavors (1a, 1b, 1c). I want my workloads to start as soon as possible on any flavor and stop pursuing other variants once the job is accommodated.

To achieve that, I use the `NoMigration` mode.

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

ditto

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: NoMigration
```

#### Story 5: Delaying Variant creation
As an admin I have two resource flavors in my cluster:
1) Most preferable: reservation
2) Less preferable: on-demand

I want to attempt scheduling on "reservation" exclusively for the first 2 hours before allowing the "on-demand" variant to activate.

To achieve this, I use explicitVariants and set createDelaySeconds for the on-demand variant.

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

ditto

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation"
        allowedResourceFlavors: ["reservation"]
      - name: "on-demand"
        allowedResourceFlavors: ["on-demand"]
        createDelaySeconds: 7200
```

#### Story 6: Limit when migration can happen
As an admin I have two resource flavors in my CQ:
1) Most preferable: reservation
2) Less preferable: on-demand

I want to avoid migrating a long-running workload if it is nearing completion. I want to constrain the migration window to the first 24 hours (86,400 seconds) of the workload's lifecycle.

To achieve this, I set maxDeleteDelaySeconds on the more preferred variant. After this time passes, the "reservation" variant is deactivated, preventing any further upgrades.

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

ditto

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation"
        allowedResourceFlavors: ["reservation"]
        maxDeleteDelaySeconds: 86400
      - name: "on-demand"
        allowedResourceFlavors: ["on-demand"]
```

#### Story 7: Workload with multiple PodSets
As an admin I have two GPU resource flavors and one CPU resource flavor in my cluster.
GPU:
1) Most preferable: reservation
2) Less preferable: on-demand

CPU:
1) default-cpu

I want to allow upgrades for the GPU portion while keeping the CPU flavor constant across variants.

> apiVersion: kueue.x-k8s.io/v1beta2

Reviewer 3: ```suggestion
```yaml
apiVersion: kueue.x-k8s.io/v1beta2
```

ditto

kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation-flavor"
        allowedResourceFlavors: ["reservation", "default-cpu"]
      - name: "on-demand-flavor"
        allowedResourceFlavors: ["on-demand", "default-cpu"]
```

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

In this section, we often refer to `ExplicitVariants`. This concept is explained in more detail in [the
Explicit Variants API section](#explicitvariants-api).

### ClusterQueue API
The ClusterQueue is extended to define the policy for concurrent attempts. This includes how to handle sibling Variants once one is admitted and how to define specific, customized Variants.

> type ClusterQueueSpec struct {

Reviewer 3: ```suggestion
```go
type ClusterQueueSpec struct {
```

ditto

    ...

    // +optional
    ConcurrentAdmissionPolicy *ConcurrentAdmissionPolicy `json:"concurrentAdmissionPolicy,omitempty"`
}

type ConcurrentAdmissionPolicy struct {
    // MigrationConstraints defines the constraints of Variants' migration
    //
    // +required
    MigrationConstraints ConcurrentAdmissionMigrationConstraints `json:"migrationConstraints"`
}

type ConcurrentAdmissionMigrationConstraints struct {
    // Mode defines the mode of Workload's migration.
    //
    // +required
    Mode ConcurrentAdmissionMigrationMode `json:"mode"`

    // MinTargetFlavor defines the minimal flavor a Workload can migrate to.
    // The order is based on the order of flavors in ClusterQueue.
    // It can only be used if the Mode is `UpgradeOnly` and `explicitVariants` is not specified.
    // If the Mode is `UpgradeOnly` and MinTargetFlavor is not specified, then there's
    // no constraints on what flavors a Workload can migrate to.
    //
    // +optional

>     MinTargetFlavor *ResourceFlavorReference `json:"minTargetFlavor,omitempty"`

Reviewer 3: I don't think that we should introduce this `minTargetFlavor` because

(1. explicitlyVariants fields (future one) could cover the sibling functionality.
(2) This is hard to understand for users (not intuitive) because there is no context that the order of flavors has a preference concept as well. 

Hence, instead of introducing minTargetFlavor, I think that we should introduce explicitVariants field.

Author: The explicitVariants doesn't cover the #story-3-reservation--homogenous-flavors also mentioned in [my comment](https://github.com/kubernetes-sigs/kueue/pull/8861#discussion_r3088290035) above. This use-case is very important for us. 

When it comes to the order of flavors to express preferences, the order is the same as in CQ's resourceFlavor configuration. It's in the same object so I believe admin will have the full context, once we provide them examples on how to use this API 

The `explicitVariants` and `minTargetFlavor` are not mutually exclusive. We plan to implement them both, it's just the story 3 is more important for us

Reviewer 3: > The explicitVariants and minTargetFlavor are not mutually exclusive. We plan to implement them both, it's just the story 3 is more important for us

Is this right? As I check the API documentation, both fields are matually exclusive, " // It can only be used if the Mode is `UpgradeOnly` and `explicitVariants` is not specified.":

```
    // MinTargetFlavor defines the minimal flavor a Workload can migrate to.
    // The order is based on the order of flavors in ClusterQueue. 
    // It can only be used if the Mode is `UpgradeOnly` and `explicitVariants` is not specified.
    // If the Mode is `UpgradeOnly` and MinTargetFlavor is not specified, then there's
    // no constraints on what flavors a Workload can migrate to.
```

Author: Sorry I wasn't clear. An admin cannot set both minTargetFlavor and explicitVariants at the same time. I meant that we can implement them both instead of choosing which to implement

Reviewer 3: Thank you! I got it

Reviewer 3: I summarized offline discussion about API in https://github.com/kubernetes-sigs/kueue/pull/8861#discussion_r3102297015

It should also activate and deactivate (via `.spec.active`) Variants based on `CreateDelaySeconds` and `MaxDeleteDelaySeconds` fields.

#### Eviction
A Variant can be evicted because of its sibling Variant during the "upgrade" process, in that case the evicted Variant
is simply deactivated.

In case of preemption by other Workloads (e.g. priority-based preemption), we reset all the Variants and treat them as if
Parent Workload has just been created - we reset the delay countdown, and activate all Variants again (beside those  with `CreateDelaySeconds`)

#### Lifecycle of a Parent and its Variants

A Parent and its Variant share very similar lifecycle:
1. Upon the submission of a job to the ClusterQueue with ConcurrentAdmission configured, a Workload with the parent label is created.
The Parent Workload doesn't take part in scheduling cycles of Kueue scheduler.
2. The Variant Controller picks up the Parent Workload and creates Variants based on the ClusterQueue configuration.
3. All Variants are picked up by the Kueue scheduler.
4. If one of the Variants is admitted, the Parent Workload gets admitted. Other Variants either get deactivated, or keep

> trying to get scheduled based on `migrationConstraints` configuration.

Reviewer 3: I couldn't understand what "keep trying to get scheduled based on `migrationConstraints` configuration" means. 

1. Why do we need to keep scheduling remaining variants even after one of the variants is admitted?
2. What migration constraints configurations for this behavior?
3. What if another variant is admitted when there is already another variant admitted?

Reviewer 3: Also, in case of "or keep trying to get scheduled based on `migrationConstraints` configuration.", what if another variant needs to preempt the already admitted variant Workload?

I meant that 

1. Create Parent Workload
2. Create 2 Variant Workloads, variant-a, and variant-b
3. variant-a is admitted, and variant-b is still on-going for the proper assignment
4. The variant-b needs to preempt variant-a for accommodation.

What happens in phase 4? Which preemption or waiting for another preemption targets?
This might be able to convert to a more principal question: Can Variant Workloads preempt each other?

Author: The general rule is that order or ResourceFlavors in CQ dictates which variants are the most favorable. This means that if more favorable variant (e.g reservation) is admitted, then less favorable variant(s)(e.g. spot) are deactivated. However if less favorable variant is admitted, more favorable ones are still pursued and if admitted, they will preempt less favorable ones and deactivate them.

Additionally we have the minTargetFlavor. This is especially useful in case where some variants are equivalent - e.g. pursuing the same type capacity in different zones. In that case we want to migrate only to e.g. reservation instead of equivalent spot which is in different zone

Author: Fully answering your questions:

> Why do we need to keep scheduling remaining variants even after one of the variants is admitted?

We would only keep scheduling **more favorable** variants, so migration can happen 

> What migration constraints configurations for this behavior?

What I meant here is the `minTargetFlavor`. If one variant is admitted all the variants below this target will be deactivated. This also accounts for future `NoMigration` mode.

> What if another variant is admitted when there is already another variant admitted?

If another variant is admitted it means it be more favorable, otherwise it keep trying to get scheduled. In that case the less favorable variant will be preempted, and the more favorable one will get admitted

> Can Variant Workloads preempt each other?

Yes, more favorable variant can preempt less favorable one

> Additionally we have the minTargetFlavor. This is especially useful in case where some variants are equivalent - e.g. pursuing the same type capacity in different zones. In that case we want to migrate only to e.g. reservation instead of equivalent spot which is in different zone

This is also mentioned as `#story-3-reservation--homogenous-flavors`

Reviewer 3: Thank you for answering my question! That sounds reasonable to me!

### Ordering Variants

The `VariantsController` leverages the existing implementation of scheduling logic in Kueue, where Workloads are sorted
by priorities and creation timestamps and then picked-up by the scheduler one per scheduling cycle (per ClusterQueue).
This means when creating Variants the controller should ensure more favorable Variants are ahead of less favorable ones in the queue.
This could be achieved by adding another dimension to our heap sort mechanism that would only
be used for sorting sibling Variants. The value of this dimension would be filled by the Variant Controller based on the order of ResourceFlavors in ClusterQueue.
Thanks to that we also have a guarantee that sibling Variants are always adjacent in the heap, which
results in Kueue scheduler picking sibling Variants one after the another, without any other Workloads in between.

### Notes/Constraints/Caveats (Optional)

#### Number of Variants

To limit the number of Variants Kueue creates based on the ClusterQueue's configuration of ResourceFlavors, we'll

> introduce the validation that limits number of ResourceGroups in ClusterQueue to 1, and the number of ResourceFlavors in that ResourceGroup to 16.

Reviewer 3: What if the possible number of variants is over 16? Will we validate such a parent Workload and deactivate it before variant creation?

Author: We can validate the number of resource flavors on the CQs that have ConcurrentAdmissionPolicy enabled, wdyt

Reviewer 3: > We can validate the number of resource flavors on the CQs that have ConcurrentAdmissionPolicy enabled, wdyt

That sounds good to me. If the ClusterQueue is invalid from the ConcurrentAdmissionPolicy PoV, the CQ creation and updating operation will be rejected by the webhook, right?

If yes, that sounds good to me.

Author: Exactly, thanks

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

#### Integration tests

In the first iteration of the feature integration tests will be added to cover following scenarios:
- Parent Workloads are not picked up by the scheduler
- Scheduler assigns flavors with respect to the flavor assignment constraints
- Variant controller creates proper number of Variant Workloads with respect to CQ's config
- Only one Variant can preempt other Workloads
- Variants are created/deleted upon the changes in CQ's config
- Workload can migrate to a target flavor if the quota is released
- Workload cannot migrate to a different flavor if the flavor is below min target

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

### Graduation Criteria

#### Alpha
In Alpha version the feature will be gated behind the `ConcurrentAdmission` feature gate.

Support for `MigrationConstraints.Mode=UpgradeOnly` and `MigrationConstraints.MinTargetFlavor` .

Integration with `BestEffortFIFO` queueing strategy.

> Introduction of `AdmissionConstraints` field.

Reviewer 3: ```suggestion
- In Alpha version the feature will be gated behind the `ConcurrentAdmission` feature gate.
- Support for `MigrationConstraints.Mode=UpgradeOnly` and `MigrationConstraints.MinTargetFlavor` .
- Integration with `BestEffortFIFO` queueing strategy.
- Introduction of `AdmissionConstraints` field.
```

Using bullet points would make this KEP easier to read.

#### Beta

Support for `MigrationConstraints.Mode=NoMigration` and `MigrationConstraints.MinTargetExplicitVariant`.

Introduction of `ExplicitVariants` functionality.

Revisit extending `ExplicitVariants` API with some additional fields.

Minimizing number of Variants issuing preemptions to only one per Parent.

Revisit the idea of [introducing WorkloadType API](#workloadtypes-api).

Positive feedback from users.

Adding/updating Kueue metrics based on users' feedback.

Revisit the [`WorkloadStatus`](#workload-status) changes.

> Revisit support for ClusterQueues with more than 1 `ResourceGroup`.

Reviewer 3: ```suggestion
- Support for `MigrationConstraints.Mode=NoMigration` and `MigrationConstraints.MinTargetExplicitVariant`.
- Introduction of `ExplicitVariants` functionality.
- Revisit extending `ExplicitVariants` API with some additional fields.
- Minimizing number of Variants issuing preemptions to only one per Parent.
- Revisit the idea of [introducing WorkloadType API](#workloadtypes-api).
- Positive feedback from users.
- Adding/updating Kueue metrics based on users' feedback.
- Revisit the [`WorkloadStatus`](#workload-status) changes.
- Revisit support for ClusterQueues with more than 1 `ResourceGroup`.
```

ditto

Reviewer 3: @PBundyra Also, could you add WorkloadSlice support here?
Or do you want to add WorkloadSlice to GA criteria?

Reviewer 3: Also, could you add FlavorFungibility support to Beta?

Reviewer 3: Also, could you add multiple preemptions to Beta?

As an admin I have two GPU resource flavors and one CPU resource flavor in my cluster.
GPU:
1) Most preferable: reservation
2) Less preferable: on-demand

CPU:
1) default-cpu

I want to allow upgrades for the GPU portion while keeping the CPU flavor constant across variants.

```yaml
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migrationConstraints:
      mode: UpgradeOnly
    explicitVariants:
      - name: "reservation-flavor"
        allowedResourceFlavors: ["reservation", "default-cpu"]
      - name: "on-demand-flavor"
        allowedResourceFlavors: ["on-demand", "default-cpu"]
```

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

In this section, we often refer to `ExplicitVariants`. This concept is explained in more detail in [the
Explicit Variants API section](#explicitvariants-api).

> ### ClusterQueue API

Reviewer 3: I deeply synced with @PBundyra offline on what we make the API.
In conclusion, we decided to introduce the following API structure to represent the API restrictions and semantics.

```yaml
spec:
  concurrentAdmissionPolicy:
    migration:
      # I am still considering the mode field name and its enum. I leave a comment for the alternatives in the thread.
      mode: UpgradeOnly | NoMigration
      constraints:
        minPreferredFlavorName: ""
        explicitVariants:              # We will revisit the `explicitVariants` API structure and field name after 0.18.
        - name: ""
          allowedResourceFlavorNames: [],
          createDelaySeconds: ""
```

Additionally, we discussed the additional `mostPreferredFlavorName` constraint for story 3 (homogeneous pattern), and the `mostPreferredFlavorName` can bypass the flavor preference order defined in `.spec.resourceGroups` at ClusterQueue while the `minPreferredFlavorName` respects the flavor preference order defined in .spec.resourceGroup.

However, we decided not to introduce `mostPreferredFlavorName` at this time because the `mostPreferredFlavorName` could be considered as a subset of `minPreferredFlavorName`, and the primary difference point is the capability to bypass flavor preference order.

The primary reason for this decision is that we always used to respect the flavor preference order during scheduling before Concurrent Admission. So, we should not introduce such a special bypassing mechanism at this time.

Reviewer 3: In addition, I'm wondering if we can refine the mode enum in the following:

```yaml
spec:
  concurrentAdmissionPolicy:
    migration:
      mode: TryPreferredFlavors | HoldFirstAdmission
```

The `TryPreferredFlavors` is an alternative to `UpgradeOnly`, and the `HoldFirstAdmission` is an alternative to `NoMigration`.

My proposal indicates the following semantics:

- `TryPreferredFlavors`: Kueue TryPreferredFlavors (tries preferred flavors) in the migration mode for the concurrent admission. Kueue doesn't stop other Workload Variants from scheduling even after one of the Workload Variants is admitted.
- `HoldFirstAdmission`: Kueue HoldFirstAdmission (holds the first admission) in the migration mode for the concurrent admission. Kueue stops other Workload Variants from scheduling after one of the Workload Variants is admitted, even if the admitted Workload Variant is not the preferred one.

@PBundyra, what do you think?

Author: SGTM, already added it. Thank you!

```yaml
kind: ClusterQueue
metadata:
  name: "cluster-queue"
spec:
  ...
  concurrentAdmissionPolicy:
    migration:
      mode: TryPreferredFlavors
      constraints:
        explicitVariants:
          - name: "reservation-flavor"
            allowedResourceFlavors: ["reservation", "default-cpu"]
          - name: "on-demand-flavor"
            allowedResourceFlavors: ["on-demand", "default-cpu"]
```

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

In this section, we often refer to `ExplicitVariants`. This concept is explained in more detail in [the
Explicit Variants API section](#explicitvariants-api).

### ClusterQueue API
The ClusterQueue is extended to define the policy for concurrent attempts. This includes how to handle sibling Variants once one is admitted and how to define specific, customized Variants.

```go
type ClusterQueueSpec struct {
    ...

>     // ConcurrenetAdmissionPolicy defines policies of the Concurrent Admission.

Reviewer 3: ```suggestion
    //ConcurrentAdmissionPolicy defines the policies of the Concurrent Admission.
```

nit: typo fixes

Author: Done

    //
    // +optional
    ConcurrentAdmissionPolicy *ConcurrentAdmissionPolicy `json:"concurrentAdmissionPolicy,omitempty"`
}

type ConcurrentAdmissionPolicy struct {
    // Migration defines the constraints of Variants' migration
    //
    // +required
    Migration ConcurrentAdmissionMigration `json:"migration"`
}

type ConcurrentAdmissionMigration struct {
    // Mode defines the mode of Workload's migration.
    //
    // +required
    Mode ConcurrentAdmissionMigrationMode `json:"mode"`

    // Constraints defines constraints on the migration. 
    //
    // +optional

>     Constraints *ConcurrentAdmissionConstraints `json:"constraints"`

Reviewer 3: ```suggestion
    Constraints *ConcurrentAdmissionConstraints `json:"constraints,omitempty"`
```

missing `omitempty` tag.

It should also activate and deactivate (via `.spec.active`) Variants based on `CreateDelaySeconds` and `MaxDeleteDelaySeconds` fields.

#### Eviction
A Variant can be evicted because of its sibling Variant during the "upgrade" process, in that case the evicted Variant
is simply deactivated.

In case of preemption by other Workloads (e.g. priority-based preemption), we reset all the Variants and treat them as if
Parent Workload has just been created - we reset the delay countdown, and activate all Variants again (beside those  with `CreateDelaySeconds`)

#### Lifecycle of a Parent and its Variants

A Parent and its Variant share very similar lifecycle:
1. Upon the submission of a job to the ClusterQueue with ConcurrentAdmission configured, a Workload with the parent label is created.
The Parent Workload doesn't take part in scheduling cycles of Kueue scheduler.
2. The Variant Controller picks up the Parent Workload and creates Variants based on the ClusterQueue configuration.
3. All Variants are picked up by the Kueue scheduler.
4. If one of the Variants is admitted, the Parent Workload gets admitted. Other Variants either get deactivated, or keep

> trying to get scheduled based on `migrationConstraints` configuration.

Reviewer 3: ```suggestion
trying to get scheduled based on `migration.constraints` configuration.
```

to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

#### Integration tests

In the first iteration of the feature integration tests will be added to cover following scenarios:
- Parent Workloads are not picked up by the scheduler
- Scheduler assigns flavors with respect to the flavor assignment constraints
- Variant controller creates proper number of Variant Workloads with respect to CQ's config
- Only one Variant can preempt other Workloads
- Variants are created/deleted upon the changes in CQ's config
- Workload can migrate to a target flavor if the quota is released
- Workload cannot migrate to a different flavor if the flavor is below min target

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

### Graduation Criteria

#### Alpha
- In Alpha version the feature will be gated behind the `ConcurrentAdmission` feature gate.

> - Support for `MigrationConstraints.Mode=TryPreferredFlavors` and `MigrationConstraints.MinPreferredFlavorName` .

Reviewer 3: ```suggestion
- Support for `migration.constraints.mode=TryPreferredFlavors` and `migration.constraints.minPreferredFlavorName` .
```

- Integration with `BestEffortFIFO` queueing strategy.

> - Introduction of `AdmissionConstraints` field.

Reviewer 3: Shouldn't we move this to Beta?
Because we will use `kueue.x-k8s.io/workload-allowed-resource-flavors` annotation for a while, right?

Author: Oh yes, will do

Author: Done

#### Beta

> - Support for `MigrationConstraints.Mode=HoldFirstAdmission` and `MigrationConstraints.MinTargetExplicitVariant`.

Reviewer 3: ```suggestion
- Support for `migration.constraints.mode=HoldFirstAdmission` and `migration.constraints.MinTargetExplicitVariant`.
```
